README.md
# `gwosc` client API

The `gwosc` package provides an interface to querying the open data
releases hosted on <https://gwosc.org> from the GEO, LIGO,
and Virgo gravitational-wave observatories.

## Release status

[![PyPI version](https://badge.fury.io/py/gwosc.svg)](http://badge.fury.io/py/gwosc)
[![Conda version](https://img.shields.io/conda/vn/conda-forge/gwosc.svg)](https://anaconda.org/conda-forge/gwosc/)  
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.1196306.svg)](https://doi.org/10.5281/zenodo.1196306)
[![License](https://img.shields.io/pypi/l/gwosc.svg)](https://choosealicense.com/licenses/mit/)
![Supported Python versions](https://img.shields.io/pypi/pyversions/gwosc.svg)

## Development status

[![Build status](https://git.ligo.org/gwosc/client/badges/main/pipeline.svg)](https://git.ligo.org/gwosc/client/-/pipelines)
![Code coverage](https://git.ligo.org/gwosc/client/badges/main/coverage.svg)
[![Documentation](https://readthedocs.org/projects/gwosc/badge/?version=latest)](https://gwosc.readthedocs.io/en/latest/?badge=latest)

## Installation

To install:

    conda install -c conda-forge gwosc

or

    pip install gwosc

## Searching for datasets

To search for available datasets (correct as of March 14 2018):

```python
>>> from gwosc import datasets
>>> datasets.find_datasets()
['GW150914', 'GW151226', 'GW170104', 'GW170608', 'GW170814', 'GW170817', 'LVT151012', 'O1', 'S5', 'S6']
>>> datasets.find_datasets(detector='V1')
['GW170814', 'GW170817']
>>> datasets.find_datasets(type='run')
['O1', 'S5', 'S6']
```

To query for the GPS time of an event dataset (or vice-versa):

```python
>>> datasets.event_gps('GW170817')
1187008882.43
>>> datasets.event_at_gps(1187008882)
'GW170817'
```

Similar queries are available for observing run datasets:

```python
>>> datasets.run_segment('O1')
(1126051217, 1137254417)
>>> datasets.run_at_gps(1135136350)  # event_gps('GW151226')
'O1'
```

## Locating data URLs by event name

You can search for remote data URLS based on the event name:

```python
>>> from gwosc.locate import get_event_urls
>>> get_event_urls('GW150914')
['https://gwosc.org/eventapi/json/GWTC-1-confident/GW150914/v3/H-H1_GWOSC_4KHZ_R1-1126259447-32.hdf5', 'https://gwosc.org/eventapi/json/GWTC-1-confident/GW150914/v3/H-H1_GWOSC_4KHZ_R1-1126257415-4096.hdf5', 'https://gwosc.org/eventapi/json/GWTC-1-confident/GW150914/v3/L-L1_GWOSC_4KHZ_R1-1126259447-32.hdf5', 'https://gwosc.org/eventapi/json/GWTC-1-confident/GW150914/v3/L-L1_GWOSC_4KHZ_R1-1126257415-4096.hdf5']
```

You can down-select the URLs using keyword arguments:

```python
>>> get_event_urls('GW150914', detector='L1', duration=32)
['https://gwosc.org/eventapi/json/GWTC-1-confident/GW150914/v3/L-L1_GWOSC_4KHZ_R1-1126259447-32.hdf5']
```

## Locating data URLs by GPS interval

You can search for remote data URLs based on the GPS time interval as
follows:

```python
>>> from gwosc.locate import get_urls
>>> get_urls('L1', 968650000, 968660000)
['https://gwosc.org/archive/data/S6/967835648/L-L1_LOSC_4_V1-968646656-4096.hdf5', 'https://gwosc.org/archive/data/S6/967835648/L-L1_LOSC_4_V1-968650752-4096.hdf5', 'https://gwosc.org/archive/data/S6/967835648/L-L1_LOSC_4_V1-968654848-4096.hdf5', 'https://gwosc.org/archive/data/S6/967835648/L-L1_LOSC_4_V1-968658944-4096.hdf5']
```

This arguments for this function are as follows

-   `detector` : the prefix of the relevant gravitational-wave
    interferometer, either `'H1'` for LIGO-Hanford, or `'L1'` for LIGO
    Livingston,
-   `start`: the GPS start time of the interval of interest
-   `end`: the GPS end time of the interval of interest

By default, this method will return the paths to HDF5 files for the 4
kHz sample-rate data, these can be specified as keyword arguments. For
full information, run

```python
>>> help(get_urls)
```

## Query for Timeline segments

You can also search for Timeline segments, based on a flag name, and a
GPS time interval as follows:

```python
>>> from gwosc.timeline import get_segments
>>> get_segments('H1_DATA', 1126051217, 1126151217)
[(1126073529, 1126114861), (1126121462, 1126123267), (1126123553, 1126126832), (1126139205, 1126139266), (1126149058, 1126151217)]
```

The output is a `list` of `(start, end)` 2-tuples which each represent a
semi-open time interval.

For documentation on what flags are available, for example for the O1
science run, see [the O1 data release page](https://gwosc.org/O1/)
(*Data Quality*).



debian/changelog
gwosc (0.7.1-1) unstable; urgency=low

  * update to 0.7.1

 -- Duncan Macleod <duncan.macleod@ligo.org>  Thu, 20 Apr 2023 10:30:00 +0100

gwosc (0.7.0-1) unstable; urgency=low

  * update to 0.7.0

 -- Duncan Macleod <duncan.macleod@ligo.org>  Mon, 10 Apr 2023 14:23:00 +0100

gwosc (0.6.1-1) unstable; urgency=low

  * update to 0.6.1

 -- Duncan Macleod <duncan.macleod@ligo.org>  Thu, 12 Aug 2021 21:37:00 +0100

gwosc (0.6.0-1) unstable; urgency=low

  * update to 0.6.0

 -- Duncan Macleod <duncan.macleod@ligo.org>  Mon, 09 Aug 2021 09:42:00 +0100

gwosc (0.5.8-1) unstable; urgency=low

  * update to 0.5.8

 -- Duncan Macleod <duncan.macleod@ligo.org>  Wed, 19 May 2021 10:41:00 +0100

gwosc (0.5.7-1) unstable; urgency=low

  * update to 0.5.7
  * add setuptools-scm and wheel build requirements
  * run tests in color

 -- Duncan Macleod <duncan.macleod@ligo.org>  Wed, 12 May 2021 12:12:00 +0100

gwosc (0.5.6-1) unstable; urgency=low

  * update to 0.5.6
  * add python3-requests-mock as a test requirement and skip the tests for
    stretch (requests-mock is too old)

 -- Duncan Macleod <duncan.macleod@ligo.org>  Thu, 27 Aug 2020 13:51:00 +0100

gwosc (0.5.5-1) unstable; urgency=low

  * update to 0.5.5

 -- Duncan Macleod <duncan.macleod@ligo.org>  Mon, 27 Jul 2020 14:23:00 +0100

gwosc (0.5.4-1) unstable; urgency=low

  * update to 0.5.4

 -- Duncan Macleod <duncan.macleod@ligo.org>  Sun, 26 Jul 2020 18:02:00 +0100

gwosc (0.5.3-1) unstable; urgency=low

  * update to 0.5.3

 -- Duncan Macleod <duncan.macleod@ligo.org>  Wed, 22 Apr 2020 17:35:00 +0000

gwosc (0.5.2-1) unstable; urgency=low

  * update to 0.5.2

 -- Duncan Macleod <duncan.macleod@ligo.org>  Wed, 18 Mar 2020 11:00:00 +0000

gwosc (0.5.1-1) unstable; urgency=low

  * update to 0.5.1

 -- Duncan Macleod <duncan.macleod@ligo.org>  Tue, 17 Mar 2020 22:34:00 +0000

gwosc (0.5.0-1) unstable; urgency=low

  * drop support for python2

 -- Duncan Macleod <duncan.macleod@ligo.org>  Tue, 17 Mar 2020 10:00:00 +0000

gwosc (0.4.3-1) unstable; urgency=low

  * [#33] fixed bugs in `gwosc.timeline`
  * [#34] added `match` keyword to `gwosc.datasets.find_datsets`

 -- Duncan Macleod <duncan.macleod@ligo.org>  Tue, 12 Mar 2019 07:19:00 +0000

gwosc (0.4.2-1) unstable; urgency=low

  * [#31] new function `gwosc.datasets.dataset_type`
  * [#32] fixed critical bug in `gwosc.locate.get_urls`

 -- Duncan Macleod <duncan.macleod@ligo.org>  Mon, 11 Mar 2019 16:39:00 +0000

gwosc (0.4.1-1) unstable; urgency=low

  * development release to include catalogue parsing

 -- Duncan Macleod <duncan.macleod@ligo.org>  Thu, 28 Feb 2019 21:18:00 +0100

gwosc (0.3.4-1) unstable; urgency=low

  * testing bug-fix release

 -- Duncan Macleod <duncan.macleod@ligo.org>  Mon, 01 Oct 2018 20:16:00 +0100

gwosc (0.3.3-1) unstable; urgency=low

  * packaging bug-fix release

 -- Duncan Macleod <duncan.macleod@ligo.org>  Mon, 09 Jul 2018 15:55:00 +0100

gwosc (0.3.1-1) unstable; urgency=low

  * development release

 -- Duncan Macleod <duncan.macleod@ligo.org>  Sat, 07 Jul 2018 14:18:00 +0100


gwosc/locate.py
# -*- coding: utf-8 -*-
# Copyright (C) Cardiff University (2018-2021)
# SPDX-License-Identifier: MIT

"""
`gwosc.locate` provides functions to determine the file URLs containing
data for a specific dataset.

You can search for remote data URLS based on the event name:

>>> from gwosc.locate import get_event_urls
>>> get_event_urls('GW150914')
['https://gwosc.org/eventapi/json/GWTC-1-confident/GW150914/v3/H-H1_GWOSC_4KHZ_R1-1126259447-32.hdf5',
 'https://gwosc.org/eventapi/json/GWTC-1-confident/GW150914/v3/H-H1_GWOSC_4KHZ_R1-1126257415-4096.hdf5',
 'https://gwosc.org/eventapi/json/GWTC-1-confident/GW150914/v3/L-L1_GWOSC_4KHZ_R1-1126259447-32.hdf5',
 'https://gwosc.org/eventapi/json/GWTC-1-confident/GW150914/v3/L-L1_GWOSC_4KHZ_R1-1126257415-4096.hdf5']

You can down-select the URLs using keyword arguments:

>>> get_event_urls('GW150914', detector='L1', duration=32)
['https://gwosc.org/eventapi/json/GWTC-1-confident/GW150914/v3/L-L1_GWOSC_4KHZ_R1-1126259447-32.hdf5']

You can search for remote data URLs based on the GPS time interval as
follows:

>>> from gwosc.locate import get_urls
>>> get_urls('L1', 968650000, 968660000)
['https://gwosc.org/archive/data/S6/967835648/L-L1_LOSC_4_V1-968646656-4096.hdf5',
 'https://gwosc.org/archive/data/S6/967835648/L-L1_LOSC_4_V1-968650752-4096.hdf5',
 'https://gwosc.org/archive/data/S6/967835648/L-L1_LOSC_4_V1-968654848-4096.hdf5',
 'https://gwosc.org/archive/data/S6/967835648/L-L1_LOSC_4_V1-968658944-4096.hdf5']

By default, this method will return the paths to HDF5 files for the 4 kHz
sample-rate data, these can be specified as keyword arguments.
For full information, see :func:`get_urls`.
"""  # noqa: E501

from . import (
    api,
    datasets,
    urls as lurls,
    utils,
)

__all__ = ['get_urls', 'get_run_urls', 'get_event_urls']
__author__ = 'Duncan Macleod <duncan.macleod@ligo.org>'


def get_urls(
        detector, start, end,
        dataset=None,
        version=None,
        sample_rate=4096,
        format='hdf5',
        host=api.DEFAULT_URL,
):
    """Fetch the URLs from GWOSC regarding a given GPS interval

    Parameters
    ----------
    detector : `str`
        the prefix of the relevant GW detector

    start : `int`
        the GPS start time of your query

    end : `int`
        the GPS end time of your query

    dataset : `str`, `None`, optional
        the name of the dataset to query, e.g. ``'GW150914'``

    version : `int`, `None`, optional
        the data-release version for the selected datasets

    sample_rate : `int`, optional, default : ``4096``
        the sampling rate (Hz) of files you want to find

    format : `str`, optional, default: ``'hdf5'``
        the file format (extension) you want to find

    host : `str`, optional
        the URL of the remote GWOSC server

    Returns
    -------
    urls : `list` of `str`
        the list of remote file URLs that contain data matching the
        relevant parameters
    """
    start = int(start)
    end = int(end)

    for dstype in ("event", "run"):
        dsets = datasets._iter_datasets(
            match=dataset,
            type=dstype,
            detector=detector,
            segment=(start, end),
            version=version,
            host=host,
        )

        for dst in dsets:
            # get URL list for this dataset
            if dstype == "run":
                urls = get_run_urls(
                    dst,
                    detector,
                    start,
                    end,
                    host=host,
                    sample_rate=sample_rate,
                    format=format,
                )
            else:
                urls = get_event_urls(
                    dst,
                    detector=detector,
                    start=start,
                    end=end,
                    format=format,
                    sample_rate=sample_rate,
                    version=version,
                    host=host,
                )

            # if full span covered, return now
            if utils.full_coverage(urls, (start, end)):
                return urls

    raise ValueError("Cannot find a GWOSC dataset for %s covering [%d, %d)"
                     % (detector, start, end))


def get_run_urls(
        run,
        detector,
        start,
        end,
        format='hdf5',
        sample_rate=4096,
        **match
):
    """Fetch the URLs from GWOSC regarding a given event

    Parameters
    ----------
    event : `str`
        the ID of the event

    detector : `str`, optional
        the detector for files you want to find

    start : `int`
        the GPS start time of your query

    end : `int`
        the GPS end time of your query

    format : `str`, optional, default: ``'hdf5'``
        the file format (extension) you want to find

    sample_rate : `int`, optional, default : ``4096``
        the sampling rate (Hz) of files you want to find

    host : `str`, optional
        the URL of the remote GWOSC server

    version : `int`, `None`, optional
        the data-release version for the selected datasets

    Returns
    -------
    urls : `list` of `str`
        the list of remote file URLs that contain data matching the
        relevant parameters
    """
    # get the URL file list for this run segment
    meta = api.fetch_run_json(
        run,
        detector,
        gpsstart=start,
        gpsend=end,
        host=match.pop('host', api.DEFAULT_URL),
    )

    # filter
    return _urls(lurls.sieve(
        meta["strain"],
        format=format,
        sampling_rate=sample_rate,
        **match
    ))


def get_event_urls(
        event,
        catalog=None,
        version=None,
        detector=None,
        start=None,
        end=None,
        format='hdf5',
        sample_rate=4096,
        host=api.DEFAULT_URL,
        **match
):
    """Fetch the URLs from GWOSC regarding a given event

    Parameters
    ----------
    event : `str`
        the ID of the event

    detector : `str`, optional
        the detector for files you want to find

    format : `str`, optional, default: ``'hdf5'``
        the file format (extension) you want to find

    sample_rate : `int`, optional, default : ``4096``
        the sampling rate (Hz) of files you want to find

    host : `str`, optional
        the URL of the remote GWOSC server

    start : `int`
        the GPS start time of your query

    end : `int`
        the GPS end time of your query

    version : `int`, `None`, optional
        the data-release version for the selected datasets

    Returns
    -------
    urls : `list` of `str`
        the list of remote file URLs that contain data matching the
        relevant parameters
    """
    # get the URL file list for this event
    meta = list(api._fetch_allevents_event_json(
        event,
        catalog=catalog,
        version=version,
        full=True,
        host=host,
    )["events"].values())[0]
    strain = meta["strain"]

    # format start and end as a segment
    if start is None and end is None:
        segment = None
    else:
        segment = (
            -float('inf') if start is None else start,
            +float('inf') if end is None else end,
        )

    # then filter on everything els
    return _urls(lurls.sieve(
        strain,
        detector=detector,
        format=format,
        sampling_rate=sample_rate,
        segment=segment,
        **match
    ))


def _urls(strain):
    return [u["url"] for u in strain]
